### PR TITLE
mgr/orchestrator: minor improvements to orch status

### PR DESCRIPTION
**Pull request by pcuzner**
The orch status command only shows the backend and
whether it was available. This has been updated to
include the paused state and the host parallelism (size
of the threadpool) to provide more info about the
orchestrator. host parallelism is only shown for cephadm

Signed-off-by: Paul Cuzner <pcuzner@redhat.com>

### CODE REVIEW

**Comment by pcuzner** [comment]
Here's an example of the output with these changes applied
<pre>
[ceph: root@maint-1 /]# ceph orch status
Backend: cephadm
Available: True
Paused: No
Host Parallelism: 10
[ceph: root@maint-1 /]# ceph orch status --format json 

{"available": true, "backend": "cephadm", "paused": false, "workers": 10}
[ceph: root@maint-1 /]# ceph orch pause                
[ceph: root@maint-1 /]# ceph orch status --format json 

{"available": true, "backend": "cephadm", "paused": true, "workers": 10}
[ceph: root@maint-1 /]# ceph orch status               
Backend: cephadm
Available: True
Paused: Yes
Host Parallelism: 10
[ceph: root@maint-1 /]# ceph orch resume
[ceph: root@maint-1 /]# ceph orch status
Backend: cephadm
Available: True
Paused: No
Host Parallelism: 10
</pre>

File: src/pybind/mgr/orchestrator/module.py [hunk 1]
```diff
@@ -1383,8 +1383,17 @@ def _status(self, format='plain'):
...
         result = {
-            "backend": o
+            "backend": o,
+            "paused": self.is_paused(),
         }
+
+        try:
+            num_workers = self.worker_pool_size()
+        except NotImplementedError:
+            pass
```
Comment (jmolmo) [inline src/pybind/mgr/orchestrator/module.py:1393]: initialize "num_workers" here

**Review by jmolmo — APPROVED** [review]

**Comment by sebastian-philipp** [comment]
I would consider the worker pool size an implementation detail of the current cephadm manager module. I mean, I have some unordered thoughts of why we might not want to expose this number to prominently:

* It doesn't really *define* the speed of mgr/cephadm alone. 
* There is a chance that we need to rework of how the module works internally. 
* [execnet](https://tracker.ceph.com/issues/44676) (the base for remoto) is in maintenance mode. We need to move on to a different ssh module at some point in the future. And by that time, we'll probably need to rethink the the worker pool again.

I'd favor, if we could keep the worker pool size somewhat limited in visibility. I don't want anyone to depend on its existence. Adding it to `ceph orch status` seems fine though. but do we really need to add it to the orchestrator interface?

**Comment by pcuzner** [comment]
@sebastian-philipp agree it's tricky. Personally, I like transparency - letting the admin understand the depth of concurrent operations helps them understand why things could be delayed and gives more insight into how the magic happens!

**Comment by sebastian-philipp** [comment]
I think this only provides an illusion of transparency. I wouldn't advice to increase the pool size. The MGR is already on its limit for bigger clusters and increasing the pool size for bigger clusters is not going to help. And making the pool smaller for smaller clusters is also not going to make a big difference. 

The more I think about it, the more I'm convinced that we should probably hide this completely in order to not irritate users.

**Comment by pcuzner** [comment]
I'm not advocating the increase of the pool size, I'm simply suggesting we help the admin understand the limits - so they can understand what the implementation limit is instead of being in the dark. I fail to see how this would irritate the anyone.

**Comment by sebastian-philipp** [comment]
> I'm not advocating the increase of the pool size, I'm simply suggesting we help the admin understand the limits - so they can understand what the implementation limit is instead of being in the dark. I fail to see how this would irritate the anyone.

We have a lot of implementation details that define the speed and scalability of cephadm and are not obvious:

* which backend is used to save cephadm's state: config-key, in-memeory
* which ssh library we're using
* time till ssh connection timeout
* if the cephadm daemon is enabled or not
* how often the mgr/prometheus module is updating it's cache
* which parts of mgr/cephadm use parallel access to hosts
* which parts of serve.py are strictly sequential (i.e. `_check_for_strays()`  on host a and `_apply_all_services` on host b
* the size of the thread pool
* the size of the ssh connection pool

I don't see why the size of the thread pool should be exposed.

**Review by mgfritch — COMMENTED** [review]
similar to https://github.com/ceph/ceph/pull/38505#discussion_r555252175, I'm not convinced this will help given the potential to become bound against the performance of the mgr ... 

but, _maybe_ it would be nice to have an internal setting (hidden?) that can "tune" the pool size during runtime if/when any of the conditions raised in https://github.com/ceph/ceph/pull/38695#issuecomment-757977171 have changed.

**Comment by pcuzner** [comment]
Maybe my view is slanted by our downstream users and ansible. Ansible has a forks parameter that clearly shows the level of parallelism that is being acted on. Since we've said goodbye to ansible with cephadm, I was hoping to at least show the level of parallelism that we can work against to help ease the transition. Yes, I know that there are heaps of internals that will also be limits - but the pool size is an obvious one and the most visible.

I'd still like to keep this in - I don't see an issue for the admin, we're not making it changeable and it just makes cephadm more relatable for those users coming from the ansible world.

**Comment by sebastian-philipp** [comment]
I think the biggest difference is that the forks in ansible are actually meaningful compared to cephadm. I think, cephadm is in many ways, very similar to Rook, pretty much opaque in how it works internally. And there is no easy way to change this. On the contrary, by adding performance improvements, like caching layers or the cephadm daemon, we're actively working towards making thread pool even more obscure and less relevant. 

On the other hand, this means we have to be very good at just doing the right thing, have a great user experience and properly informing users.

**Comment by tserong** [comment]
+1 for showing the paused state. As for the worker pool, IMO it'd make sense to show it if it were possible to change the setting, but AFAICT it's hard-coded to 10. Assuming that's the case couldn't we just document that somewhere?

**Comment by pcuzner** [comment]
@tserong  agree - the paused state is a no brainer..as far as parallelism is concerned, yes document naturally - but an rtfm just to understand the parallelism seems harsh to me - IMO we have to make things more obvious and intuitive to make the platform more usable. Help the admin immediately understand rather than hope they've read the manual cover to cover or feel comfortable with diving in to the code base.

**Comment by jmolmo** [comment]
@sebastian-philipp :
I think this is not harmful at all for anybody and provides more information about what is the current configuration. Why not to show this information , there is no "contract" with nobody about what should me shown or not when you use the "status" command, so we are not compromising "future" modifications...

Providing information without "flooding" is always positive. +1 to merge this.

**Comment by pcuzner** [comment]
@sebastian-philipp the feedback from the consultants over night is that they see value in the setting but suggestputting it inside a "detail" subparameter like ceph health detail. So it's easy to find if you need it. Would that work for you?

**Comment by pcuzner** [comment]
@sebastian-philipp the latest patch removes the worker_pool_size from _interface and uses the available module as discussed. I've added the worker pool size to a --detail option on orch status too, so the threadpool size it's there if you want deeper info.

minor update to docs too to show the --detail option is there

**Comment by pcuzner** [comment]
jenkins test make check

**Comment by pcuzner** [comment]
jenkins test make check

**Comment by pcuzner** [comment]
jenkins test make check

**Comment by pcuzner** [comment]
jenkins retest this please

File: src/pybind/mgr/cephadm/module.py [hunk 2]
```diff
@@ -683,7 +686,8 @@ def available(self) -> Tuple[bool, str]:
             return ok, err
         if not self.ssh_key or not self.ssh_pub:
             return False, 'SSH keys not set. Use `ceph cephadm set-priv-key` and `ceph cephadm set-pub-key` or `ceph cephadm generate-key`'
-        return True, ''
+
+        return True, str(self._worker_pool._processes)  # type: ignore
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py:690]:
```suggestion
        ret = {
            'workers': self._worker_pool._processes,
            'is_paused': self.paused
        }
        return True, '', ret
```

File: src/pybind/mgr/orchestrator/module.py [hunk 3]
```diff
@@ -1343,29 +1343,39 @@ def _cancel(self) -> HandleCommandResult:
...
+    def _status(self,
+                detail: bool = False,
+                format: Format = Format.plain) -> HandleCommandResult:
         """Report configured backend and its status"""
         o = self._select_orchestrator()
         if o is None:
             raise NoOrchestrator()
 
-        avail, why = self.available()
+        avail, msg = self.available()
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/module.py:1354]:
```suggestion
        avail, why, details = self.available()
```

File: src/pybind/mgr/orchestrator/module.py [hunk 4]
```diff
@@ -1343,29 +1343,39 @@ def _cancel(self) -> HandleCommandResult:
...
+
         if avail is not None:
             result['available'] = avail
-            if not avail:
-                result['reason'] = why
+            if avail:
+                if o == "cephadm" and detail:
+                    result['workers'] = msg
+            else:
+                result['reason'] = msg
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/module.py:1366]:
```suggestion
            if avail:
                result.update(details)
            else:
                result['reason'] = why
```

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

**Comment by pcuzner** [comment]
@sebastian-philipp You've changed the signature of the function - which is what I was avoiding. The 2nd return value was always a str and unused - so my approach just expanded on what was already there. This keeps it consistent with the available function in the rook module. Could you elaborate on why your approach is preferred

**Comment by sebastian-philipp** [comment]
> Could you elaborate on why your approach is preferred

Your code introduces two new ways of passing data from cephadm to mgr/orch:

* `paused` is queried via `is_paused`
* The thread count is funneled through the `why` result of `available()`

Also, `str(self._worker_pool._processes)` doesn't properly type check, thus you're disabling the type checker. 

I'd like to avoid to introduce potential regressions similar to #38995 .

**Comment by pcuzner** [comment]
@sebastian-philipp and my comment about the signature changing from 2 to 3 and the potential valueerror when rook not  cephadm is in charge?

**Comment by sebastian-philipp** [comment]
> @sebastian-philipp and my comment about the signature changing from 2 to 3 and the potential valueerror when rook not cephadm is in charge?

no worries, #39131 is going to verify that the return values will be correct in mgr/rook

**Comment by tserong** [comment]
> @sebastian-philipp You've changed the signature of the function - which is what I was avoiding. The 2nd return value was always a str and unused - so my approach just expanded on what was already there. This keeps it consistent with the available function in the rook module. Could you elaborate on why your approach is preferred

Ah! I now understand the reason for changing the method signature. In the `(bool, str)` tuple, the bool indicates availability, and the str is intended to be a (free form) error message if the orchestrator isn't available for some reason. Overloading the str in the success case to state the number of workers seems somewhat arbitrary, and potentially confusing/tricky if we want to add more data in future. If we want to return extra data from the available method, I think it makes things clearer to add a third value to the tuple, containing those details in some structured form, as @sebastian-philipp suggested. This also makes it extensible in future.

**Comment by pcuzner** [comment]
but that's why I changed the str to be msg...not "why" - so in a OK case we can pass back a msg and in a fail case the msg is the error - which also means the impact to rook is non-existent too.

anyway, in the interests of time (this was submitted on 23rd dec) I'll make the changes and include the change to rook's "available" function

**Comment by sebastian-philipp** [comment]
jenkins test make check

**Comment by alfonsomthd** [comment]
CEPH API tests failed:
```
2021-02-04 09:20:27,140.140 INFO:__main__:======================================================================
2021-02-04 09:20:27,140.140 INFO:__main__:ERROR: test_host_daemons (tasks.mgr.dashboard.test_host.HostControllerTest)
2021-02-04 09:20:27,140.140 INFO:__main__:
2021-02-04 09:20:27,140.140 INFO:__main__:----------------------------------------------------------------------
2021-02-04 09:20:27,140.140 INFO:__main__:Traceback (most recent call last):
2021-02-04 09:20:27,140.140 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/mgr/dashboard/test_host.py", line 92, in test_host_daemons
2021-02-04 09:20:27,140.140 INFO:__main__:    hosts = [host['hostname'] for host in hosts if host['hostname'] != '']
2021-02-04 09:20:27,140.140 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/mgr/dashboard/test_host.py", line 92, in <listcomp>
2021-02-04 09:20:27,140.140 INFO:__main__:    hosts = [host['hostname'] for host in hosts if host['hostname'] != '']
2021-02-04 09:20:27,140.140 INFO:__main__:TypeError: string indices must be integers
2021-02-04 09:20:27,140.140 INFO:__main__:
```

**Comment by pcuzner** [comment]
jenkins test api

**Comment by pcuzner** [comment]
jenkins retest this please

**Comment by pcuzner** [comment]
@alfonsomthd test_orchestrator now updated to reflect the cephadm, orchestrator and rook change  - so now the API checks pass!

File: src/pybind/mgr/cephadm/module.py (hunk 2 above)
Comment (pcuzner) [reply src/pybind/mgr/cephadm/module.py]: @sebastian-philipp please review again once you get the chance. The changes requested are in and unfortunately impacted base class, cephadm, rook, the test_orchestrator and the dashboard...but we're there now.

Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/module.py]: yeah, exactly. lgtm!

File: src/pybind/mgr/orchestrator/module.py [hunk 5]
```diff
@@ -1343,29 +1343,36 @@ def _cancel(self) -> HandleCommandResult:
...
-                output += "\nAvailable: {0}".format(result['available'])
-                if 'reason' in result:
-                    output += ' ({0})'.format(result['reason'])
+            output += f"\nAvailable: {'Yes' if result['available'] else 'No'}"
+            if 'reason' in result:
+                output += ' ({0})'.format(result['reason'])
+            if 'paused' in result:
+                output += f"\nPaused: {'Yes' if result['paused'] else 'No'}"
+            if 'workers' in result and detail:
+                output += f"\nHost Parallelism: {result['workers']}"
```
Comment (epuertat) [inline src/pybind/mgr/orchestrator/module.py:1375]: That's why I like the new Python 3 f-string thing:
```py
output = f"""\
Backend: {result['backend']}
Available: {}
Reason: {}
Paused: {}
Host Parallelism: {}"""
```

**Review by epuertat — APPROVED** [review]
Looks good to me!

**Review by alfonsomthd — APPROVED** [review]

**Review by nizamial09 — APPROVED** [review]

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-02-05_14:59:08-rados-wip-kefu-testing-2021-02-05-1809-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-02-06_12:11:31-rados-wip-kefu-testing-2021-02-06-1838-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-02-07_14:19:29-rados-wip-kefu-testing-2021-02-07-2019-distro-basic-smithi/

reran the failed tests on master

- https://pulpito.ceph.com/kchai-2021-02-08_02:14:21-rados-wip-kefu2-testing-2021-02-08-0023-distro-basic-smithi/

e2e failures tracked by

- https://tracker.ceph.com/issues/49205

**Review by sebastian-philipp — APPROVED** [review]

**Comment by sebastian-philipp** [comment]
commit 4588ad184c4464560c1d5c9b2f5d03da41b2eebb is making the backport painful right now

**Pull request merged**